logic/tool/ExpressionCompiler.java

```diff
@@ -0,0 +1,150 @@
+import com.vaticle.typedb.core.pattern.variable.ThingVariable;
+import com.vaticle.typedb.core.pattern.variable.Variable;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+import com.vaticle.typedb.core.traversal.expression.Expression;
+import com.vaticle.typedb.core.traversal.expression.ExpressionFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.AMBIGUOUS_ARGUMENT_TYPE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.FUNCTION_NOT_FOUND;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.OPERATION_NOT_FOUND;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Pattern.VALUE_ASSIGNMENT_CYCLE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Pattern.VALUE_VARIABLE_UNASSIGNED;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class ExpressionCompiler {
+
+    private final GraphManager graphMgr;
+
+    public ExpressionCompiler(GraphManager graphMgr) {
+        this.graphMgr = graphMgr;
+    }
+
+    public void compileExpressions(Disjunction disjunction) {
+        List<ValueVariable> valueVariables = iterate(disjunction.conjunctions())
+                .flatMap(conjunction -> iterate(conjunction.variables()))
+                .filter(Variable::isValue).map(Variable::asValue).toList();
+
+        iterate(valueVariables).filter(v -> v.assignment() == null).forEachRemaining(v -> {
+            throw TypeDBException.of(VALUE_VARIABLE_UNASSIGNED, v);
+        });
+        valueVariables.forEach(v -> validateAcyclicAssignments(v, new HashSet<>(), new Stack<>()));
+        valueVariables.forEach(v -> compileExpression(v.asValue()));
+    }
+
+    private void validateAcyclicAssignments(ValueVariable var, Set<ValueVariable> validated, Stack<ValueVariable> stack) {
+        if (validated.contains(var)) return;
+
+        if (stack.contains(var)) {
+            stack.push(var);
+            throw TypeDBException.of(VALUE_ASSIGNMENT_CYCLE, stack);
+        }
+
+        stack.push(var);
+        iterate(var.assignment().arguments()).filter(Variable::isValue).forEachRemaining(v -> validateAcyclicAssignments(v.asValue(), validated, stack));
+        ValueVariable poppedAt = stack.pop();
+        assert poppedAt == var;
+        validated.add(var);
+    }
+
+    private Encoding.ValueType<?> compileExpression(ValueVariable v) {
+        if (v.assignment().valueType() == null) {
+            Map<Identifier, Encoding.ValueType<?>> argTypes = new HashMap<>();
+            // No circular dependencies -> No infinite recursion
+            iterate(v.assignment().arguments()).forEachRemaining(argument -> {
```

Comment:
fyi java Collections have a built-in `forEach` that mean you don't have to use a iterator and `forEachRemaining` explicitly:
```
v.assignment().arguments().forEach(argument -> ...)
```

Reply:
might be ever so slightly more performant